File: src/main/java/net/etum/shattered/gui/builder/ItemBuilder.java
package net.etum.shattered.gui.builder;

import org.bukkit.Material;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.inventory.ItemFlag;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.inventory.meta.SkullMeta;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class ItemBuilder {
    private final ItemStack item;
    private final ItemMeta meta;

    public ItemBuilder(Material material) {
        if (material == null) {
            throw new IllegalArgumentException("Material cannot be null");
        }
        this.item = new ItemStack(material);
        this.meta = item.getItemMeta();
    }

    public ItemBuilder(Material material, int amount) {
        if (material == null) {
            throw new IllegalArgumentException("Material cannot be null");
        }
        if (amount <= 0) {
            throw new IllegalArgumentException("Amount must be greater than zero");
        }
        this.item = new ItemStack(material, amount);
        this.meta = item.getItemMeta();
    }

    public ItemBuilder setName(String name) {
        meta.setDisplayName(name);
        return this;
    }

    public ItemBuilder addLore(String... lines) {
        List<String> lore = meta.hasLore() ? meta.getLore() : new ArrayList<>();
        for (String line : lines) {
            lore.add(line);
        }
        meta.setLore(lore);
        return this;
    }

    public ItemBuilder addEnchant(Enchantment enchantment, int level) {
        if (enchantment == null) {
            throw new IllegalArgumentException("Enchantment cannot be null");
        }
        meta.addEnchant(enchantment, level, true);
        return this;
    }

    public ItemBuilder addEnchants(Map<Enchantment, Integer> enchantments) {
        for (Map.Entry<Enchantment, Integer> entry : enchantments.entrySet()) {
            meta.addEnchant(entry.getKey(), entry.getValue(), true);
        }
        return this;
    }

    public ItemBuilder addItemFlag(ItemFlag flag) {
        if (flag == null) {
            throw new IllegalArgumentException("ItemFlag cannot be null");
        }
        meta.addItemFlags(flag);
        return this;
    }

    public ItemBuilder addItemFlags(ItemFlag... flags) {
        for (ItemFlag flag : flags) {
            if (flag == null) {
                throw new IllegalArgumentException("ItemFlag cannot be null");
            }
            meta.addItemFlags(flag);
        }
        return this;
    }

    public ItemBuilder hideAllFlags() {
        meta.addItemFlags(ItemFlag.values());
        return this;
    }

    public ItemBuilder setUnbreakable(boolean unbreakable) {
        meta.setUnbreakable(unbreakable);
        return this;
    }

    public ItemBuilder setDurability(int durability) {
        item.setDurability((short) durability);
        return this;
    }

    public ItemBuilder setAmount(int amount) {
        if (amount <= 0) {
            throw new IllegalArgumentException("Amount must be greater than zero");
        }
        item.setAmount(amount);
        return this;
    }

    public void setSkullOwner(String owner) {
        if (item.getType() != Material.PLAYER_HEAD) {
            throw new IllegalStateException("Item is not a player head");
        }
        SkullMeta skullMeta = (SkullMeta) meta;
        skullMeta.setOwner(owner);
    }

    public ItemBuilder setCustomModelData(int data) {
        meta.setCustomModelData(data);
        return this;
    }

    public ItemStack build() {
        item.setItemMeta(meta);
        return item;
    }

    public ItemBuilder clone() {
        ItemBuilder clone = new ItemBuilder(item.getType(), item.getAmount());
        clone.meta.setDisplayName(meta.getDisplayName());
        clone.meta.setLore(meta.getLore());
        clone.meta.setUnbreakable(meta.isUnbreakable());
        for (ItemFlag flag : meta.getItemFlags()) {
            clone.meta.addItemFlags(flag);
        }
        clone.meta.setCustomModelData(meta.getCustomModelData());
        return clone;
    }

    public static ItemBuilder fromItemStack(ItemStack itemStack) {
        if (itemStack == null) {
            throw new IllegalArgumentException("ItemStack cannot be null");
        }
        ItemBuilder builder = new ItemBuilder(itemStack.getType(), itemStack.getAmount());
        ItemMeta meta = itemStack.getItemMeta();
        if (meta != null) {
            builder.meta.setDisplayName(meta.getDisplayName());
            builder.meta.setLore(meta.getLore());
            builder.meta.setUnbreakable(meta.isUnbreakable());
            for (ItemFlag flag : meta.getItemFlags()) {
                builder.meta.addItemFlags(flag);
            }
            builder.meta.setCustomModelData(meta.getCustomModelData());
        }
        return builder;
    }
}

File: src/main/java/net/etum/shattered/gui/builder/ItemBuilder.kt
package net.etum.shattered.gui.builder

import org.bukkit.Material
import org.bukkit.enchantments.Enchantment
import org.bukkit.inventory.ItemFlag
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.ItemMeta
import org.bukkit.inventory.meta.SkullMeta

class ItemBuilder {
    private val item: ItemStack
    private val meta: ItemMeta

    constructor(material: Material?) {
        requireNotNull(material) { "Material cannot be null" }
        this.item = ItemStack(material)
        this.meta = item.itemMeta
    }

    constructor(material: Material?, amount: Int) {
        requireNotNull(material) { "Material cannot be null" }
        require(amount > 0) { "Amount must be greater than zero" }
        this.item = ItemStack(material, amount)
        this.meta = item.itemMeta
    }

    fun setName(name: String?): ItemBuilder {
        meta.setDisplayName(name)
        return this
    }

    fun addLore(vararg lines: String): ItemBuilder {
        val lore = if (meta.hasLore()) meta.lore else ArrayList()
        for (line in lines) {
            lore!!.add(line)
        }
        meta.lore = lore
        return this
    }

    fun addEnchant(enchantment: Enchantment?, level: Int): ItemBuilder {
        requireNotNull(enchantment) { "Enchantment cannot be null" }
        meta.addEnchant(enchantment, level, true)
        return this
    }

    fun addEnchants(enchantments: Map<Enchantment?, Int?>): ItemBuilder {
        for ((key, value) in enchantments) {
            meta.addEnchant(key!!, value!!, true)
        }
        return this
    }

    fun addItemFlag(flag: ItemFlag?): ItemBuilder {
        requireNotNull(flag) { "ItemFlag cannot be null" }
        meta.addItemFlags(flag)
        return this
    }

    fun addItemFlags(vararg flags: ItemFlag?): ItemBuilder {
        for (flag in flags) {
            requireNotNull(flag) { "ItemFlag cannot be null" }
            meta.addItemFlags(flag)
        }
        return this
    }

    fun hideAllFlags(): ItemBuilder {
        meta.addItemFlags(*ItemFlag.entries.toTypedArray())
        return this
    }

    fun setUnbreakable(unbreakable: Boolean): ItemBuilder {
        meta.isUnbreakable = unbreakable
        return this
    }

    fun setDurability(durability: Int): ItemBuilder {
        item.durability = durability.toShort()
        return this
    }

    fun setAmount(amount: Int): ItemBuilder {
        require(amount > 0) { "Amount must be greater than zero" }
        item.amount = amount
        return this
    }

    fun setSkullOwner(owner: String?) {
        check(item.type == Material.PLAYER_HEAD) { "Item is not a player head" }
        val skullMeta = meta as SkullMeta
        skullMeta.setOwner(owner)
    }

    fun setCustomModelData(data: Int): ItemBuilder {
        meta.setCustomModelData(data)
        return this
    }

    fun build(): ItemStack {
        item.setItemMeta(meta)
        return item
    }

    fun clone(): ItemBuilder {
        val clone = ItemBuilder(item.type, item.amount)
        clone.meta.setDisplayName(meta.displayName)
        clone.meta.lore = meta.lore
        clone.meta.isUnbreakable = meta.isUnbreakable
        for (flag in meta.itemFlags) {
            clone.meta.addItemFlags(flag!!)
        }
        clone.meta.setCustomModelData(meta.customModelData)
        return clone
    }

    companion object {
        fun fromItemStack(itemStack: ItemStack?): ItemBuilder {
            requireNotNull(itemStack) { "ItemStack cannot be null" }
            val builder = ItemBuilder(itemStack.type, itemStack.amount)
            val meta = itemStack.itemMeta
            if (meta != null) {
                builder.meta.setDisplayName(meta.displayName)
                builder.meta.lore = meta.lore
                builder.meta.isUnbreakable = meta.isUnbreakable
                for (flag in meta.itemFlags) {
                    builder.meta.addItemFlags(flag!!)
                }
                builder.meta.setCustomModelData(meta.customModelData)
            }
            return builder
        }
    }
}